src/Development/IDE/LSP/Outline.hs

```diff
@@ -0,0 +1,194 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+#include "ghc-api-version.h"
+
+module Development.IDE.LSP.Outline
+  ( setHandlersOutline
+  )
+where
+
+import qualified Language.Haskell.LSP.Core     as LSP
+import           Language.Haskell.LSP.Messages
+import           Language.Haskell.LSP.Types
+import           Data.Functor
+import           Data.Generics
+import           Data.Maybe
+import           Data.Text                      ( Text
+                                                , pack
+                                                )
+import qualified Data.Text                     as T
+import           Development.IDE.Core.Rules
+import           Development.IDE.Core.Shake
+import           Development.IDE.GHC.Compat
+import           Development.IDE.GHC.Error      ( srcSpanToRange )
+import           Development.IDE.LSP.Server
+import           Development.IDE.Types.Location
+import           Outputable                     ( Outputable
+                                                , ppr
+                                                , showSDocUnsafe
+                                                )
+
+setHandlersOutline :: PartialHandlers
+setHandlersOutline = PartialHandlers $ \WithMessage {..} x -> return x
+  { LSP.documentSymbolHandler = withResponse RspDocumentSymbols moduleOutline
+  }
+
+moduleOutline
+  :: LSP.LspFuncs () -> IdeState -> DocumentSymbolParams -> IO DSResult
+moduleOutline _lsp ideState DocumentSymbolParams { _textDocument = TextDocumentIdentifier uri }
+  = case uriToFilePath uri of
+    Just (toNormalizedFilePath -> fp) -> do
+      mb_decls <- runAction ideState $ use GetParsedModule fp
+      pure $ case mb_decls of
+        Nothing -> DSDocumentSymbols (List [])
+        Just (ParsedModule { pm_parsed_source = L _ltop HsModule { hsmodName, hsmodDecls, hsmodImports } })
+          -> let
+               declSymbols  = mapMaybe documentSymbolForDecl hsmodDecls
+               moduleSymbol = hsmodName <&> \(L l m) ->
+                 (defDocumentSymbol l :: DocumentSymbol)
+                   { _name  = pprText m
+                   , _kind  = SkFile
+                   , _range = Range (Position 0 0) (Position 10000000 0) -- _ltop is 0 0 0 0
```

Comment:
```suggestion
                   , _range = Range (Position 0 0) (Position maxBound 0) -- _ltop is 0 0 0 0
```
I'm guessing 10000000 is just meant to be larger than any actual file. Would `maxBound` be more meaningful?

Reply:
Good point, thanks!